source/fuzz/transformation_add_read_modify_write_atomic_instruction.cpp

```diff
@@ -0,0 +1,487 @@
+        uint32_t comparator_id,
+        const protobufs::InstructionDescriptor& instruction_to_insert_before) {
+  message_.set_fresh_id(fresh_id);
+  message_.set_pointer_id(pointer_id);
+  message_.set_opcode(opcode);
+  message_.set_memory_scope_id(memory_scope_id);
+  message_.set_memory_semantics_id_1(memory_semantics_id_1);
+  message_.set_memory_semantics_id_2(memory_semantics_id_2);
+  message_.set_value_id(value_id);
+  message_.set_comparator_id(comparator_id);
+
+  *message_.mutable_instruction_to_insert_before() =
+      instruction_to_insert_before;
+}
+
+bool TransformationAddReadModifyWriteAtomicInstruction::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // The result id must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // The pointer must exist and have a type.
+  auto pointer = ir_context->get_def_use_mgr()->GetDef(message_.pointer_id());
+  if (!pointer || !pointer->type_id()) {
+    return false;
+  }
+
+  // Read-Modify-Write atomic instructions are valid only.
+  if (pointer->opcode() == SpvOpAtomicLoad ||
+      pointer->opcode() == SpvOpAtomicStore) {
+    return false;
+  }
+
+  // The type must indeed be a pointer type.
+  auto pointer_type = ir_context->get_def_use_mgr()->GetDef(pointer->type_id());
+  assert(pointer_type && "Type id must be defined.");
+  if (pointer_type->opcode() != SpvOpTypePointer) {
+    return false;
+  }
+  // We do not want to allow loading from null or undefined pointers, as it is
+  // not clear how punishing the consequences of doing so are from a semantics
+  // point of view.
+  switch (pointer->opcode()) {
+    case SpvOpConstantNull:
+    case SpvOpUndef:
+      return false;
+    default:
+      break;
+  }
+
+  // Determine which instruction we should be inserting before.
+  auto insert_before =
+      FindInstruction(message_.instruction_to_insert_before(), ir_context);
+  // It must exist, ...
+  if (!insert_before) {
+    return false;
+  }
+
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(pointer->opcode(),
+                                                    insert_before)) {
+    return false;
+  }
+
+  // The block we are inserting into needs to be dead, or else the pointee type
+  // of the pointer needs to be irrelevant.
+  if (!transformation_context.GetFactManager()->BlockIsDead(
+          ir_context->get_instr_block(insert_before)->id()) &&
+      !transformation_context.GetFactManager()->PointeeValueIsIrrelevant(
+          message_.pointer_id())) {
+    return false;
+  }
+
+  // Instruction must RMW be atomic instruction, Id must be suitable for
+  // instruction, Check the validity of (Value, second memory semantics id,
+  // comparator id) for instruction if found.
+  if (!IsOperandIdsValidForInstruction(pointer->opcode(), ir_context,
+                                       insert_before, pointer_type)) {
+    return false;
+  }
```

Comment:
This has a similar issue to your changing memory semantics PR. Please update this PR similarly. We don't need to check the opcode in IsApplicable because it is part of the protobuf; we know what it is; it does not change depending on ir_context.

Reply:
I edited the comment above.

Reply:
OK, I think I see what is happening now. I think this can be made a bit simpler. I can think of two ways. 

## Approach 1

Check each protobuf field in order:

* memory_scope_id
* memory_semantics_id_1
* memory_semantics_id_2
* value_id
* comparator_id

In some cases (e.g. `memory_scope_id`), you can just always do the check. In other cases (e.g. `memory_semantics_id_2`), you can call a function like `IsMemorySemanticsId2Valid`. But modify `IsMemorySemanticsId2Valid` so that, at the start, it returns true if the opcode is one of the atomic opcodes that we support, but does NOT use the memory_semantics_id_2 field. E.g. if the opcode is `OpAtomicExchange`, this function will just immediately return true, and so the `memory_semantics_id_2` field will not actually be checked.

## Approach 2

Add a switch statement here with all opcodes that we consider, with three categories:

* Opcodes like `OpAtomicExchange` that use value_id but no comparator.
* Opcodes like `OpAtomicCompareExchange` that use value_id and the comparator.
* Opcodes like `OpAtomicIIncrement` that do not use value_id nor comparator.

In each case, call a function that does all the necessary checks for each category.

Reply:
Thanks, I will following the second approach probably.

Reply:
Done.